common/fs/src/tail.rs

```diff
@@ -79,330 +83,396 @@ pub struct Tailer {
+                                    && event_idx < (prev_event_idx + EVENT_STREAM_BUFFER_COUNT)
                                 {
-                                    // We've already processed this event, skip tailing
-                                    if previous_event_time >= event_time
-                                        && event_idx < (prev_event_idx + EVENT_STREAM_BUFFER_COUNT)
-                                    {
-                                        debug!("skipping already processed events");
-                                        Metrics::fs().increment_writes();
-                                        return None;
-                                    }
+                                    debug!("skipping already processed events");
+                                    Metrics::fs().increment_writes();
+                                    return None;
                                 }
+                            }
 
-                                let line = Tailer::handle_event(
-                                    event,
-                                    initial_offsets,
-                                    lookback_config,
-                                    fs.lock().await.deref(),
-                                )
-                                .await;
+                            let line = handle_event(
+                                event,
+                                initial_offsets,
+                                lookback_config,
+                                fs.lock().await.deref(),
+                            )
+                            .await;
 
-                                let line = line.map(|option_val| option_val.map(Ok).right_stream());
+                            let line = line.map(|option_val| option_val.map(Ok).right_stream());
 
-                                if let Some((key, _)) = key_and_previous_event_time {
-                                    let mut event_times = event_times.lock().await;
-                                    let new_event_time = chrono::offset::Utc::now();
-                                    event_times.insert(key, (event_idx, new_event_time));
-                                }
-                                line
+                            if let Some((key, _)) = key_and_previous_event_time {
+                                let mut event_times = event_times.lock().await;
+                                let new_event_time = chrono::offset::Utc::now();
+                                event_times.insert(key, (event_idx, new_event_time));
                             }
+
+                            line
                         }
                     }
                 }
-            })
-            .filter_map(|x| async move { x })
-            .flatten())
+            }
+        })
+        .filter_map(|x| async move { x })
+        .flatten())
+}
+
+impl Tailer {
+    /// Creates new instance of Tailer
+    pub fn new(
```

Comment:
The `Tailer` is not just a state struct.

Reply:
not or now?

Reply:
I meant to say now. Just too excited that it finally compiles.